src/software/backend/simulation/simulator_robot_test.cpp

```diff
@@ -3,39 +3,688 @@
     }
+
+    EXPECT_LT(simulator_ball->velocity().length(), 0.25);
+}
+
+TEST_F(SimulatorRobotTest, test_robot_kicks_ball_with_autokick_enabled)
+{
+    Robot robot(0, Point(0, 0), Vector(0, 0), Angle::zero(), AngularVelocity::zero(),
+                Timestamp::fromSeconds(0));
+    Ball ball(Point(0.25, 0), Vector(-1, 0), Timestamp::fromSeconds(0));
+    auto [world, simulator_robot, simulator_ball] = createWorld(robot, ball);
+
+    EXPECT_LT((simulator_ball->velocity() - Vector(-1, 0)).length(), 0.001);
+
+    simulator_robot->enableAutokick(5);
+
+    // Simulate for 1/2 second
+    for (unsigned int i = 0; i < 30; i++)
+    {
+        world->stepSimulation(Duration::fromSeconds(1.0 / 60.0));
+    }
+
+    EXPECT_LT((simulator_ball->velocity() - Vector(4, 0)).length(), 0.001);
+}
+
+TEST_F(SimulatorRobotTest,
+       test_robot_chips_ball_with_autochip_enabled_and_ball_does_not_land_in_obstacle)
+{
+    Robot robot(0, Point(0, 0), Vector(0, 0), Angle::zero(), AngularVelocity::zero(),
+                Timestamp::fromSeconds(0));
+    Ball ball(Point(0.15, 0), Vector(-0.25, 0), Timestamp::fromSeconds(0));
+    std::vector<Point> enemy_robot_positions = {Point(0.4, 0), Point(1.9, 0)};
+    auto [world, simulator_robot, simulator_ball] =
+        createWorldWithEnemyRobots(robot, ball, enemy_robot_positions);
+
+    EXPECT_LT((simulator_ball->velocity() - Vector(-0.25, 0)).length(), 0.001);
+
+    simulator_robot->enableAutochip(2);
+
+    // Simulate for 1.5 seconds
+    for (unsigned int i = 0; i < 90; i++)
+    {
+        world->stepSimulation(Duration::fromSeconds(1.0 / 60.0));
+    }
+
+    EXPECT_GT(simulator_ball->velocity().x(), 1);
+    EXPECT_GT(simulator_ball->position().x(), 2.0);
+}
+
+TEST_F(SimulatorRobotTest,
+       test_robot_chips_ball_with_autochip_enabled_and_ball_lands_on_obstacle)
+{
+    Robot robot(0, Point(0, 0), Vector(0, 0), Angle::zero(), AngularVelocity::zero(),
+                Timestamp::fromSeconds(0));
+    Ball ball(Point(0.15, 0), Vector(-0.25, 0), Timestamp::fromSeconds(0));
+    // Add an enemy robot right where the ball will land
+    std::vector<Point> enemy_robot_positions = {Point(2.0, 0)};
+    auto [world, simulator_robot, simulator_ball] =
+        createWorldWithEnemyRobots(robot, ball, enemy_robot_positions);
+
+    EXPECT_LT((simulator_ball->velocity() - Vector(-0.25, 0)).length(), 0.001);
+
+    simulator_robot->enableAutochip(2);
+
+    // Simulate for 1.5 seconds
+    for (unsigned int i = 0; i < 90; i++)
+    {
+        world->stepSimulation(Duration::fromSeconds(1.0 / 60.0));
+    }
+
+    // Check the ball's velocity is very roughly within a reasonable range. We don't care
+    // exactly what it is, just that it hasn't changed an impossible amount by
+    // getting caught inside the robot it landed in and then squeezed out
+    EXPECT_GT(simulator_ball->velocity().x(), 1);
+    EXPECT_LT(simulator_ball->velocity().x(), 3);
+    EXPECT_GT(simulator_ball->position().x(), 2.0);
```

Comment:
So this is kinda undefined behavior? If so, why aren't we just asserting the position is within a general radius and the velocity magnitude is within an acceptable amount? Is there an implicit assumption about where the ball will go?

Reply:
It's not undefined. We specifically implement the `PhysicsBall` and `ContactListener` so that if a ball that is chipped "lands" inside an object, that the ball simply continues to move without collisions enabled until it is no longer touching an object. This way we simulate the ball landing on top of something and rolling off.

This test is to make sure we don't run into potentially undefined behavior where if a ball suddenly had its collisions enabled while inside a solid object, it may be ejected in a different direction and/or and some high speed.

I've clarified the comment in the test. Please lmk if this makes sense.

Reply:
Ah, I see. Would be good to explicitly state "if the ball lands in an obstacle, collisions with the ball are disabled until the ball is outside the obstacle" or something similar as well.

Reply:
added comment

Reply:
Sorry, where?

Reply:
pushed, now should show up